WellState FIBO: Return Segment Vectors from report()

bska opened the pull request:
This pull request extends class `WellStateFullyImplictBlackoil` to report segment-related quantities as `Opm::data::Segment` objects (included in `Opm::data::WellRates` objects).  All wells have at least a top
segment in the context of WellState FIBO, so there is a meaningful value to report for each well.

We put the extraction of segment-related quantities into a new helper function
```
WellStateFullyImplicitBlackoil::reportSegmentResults()
```
to avoid cluttering up the body of `report()` more than absolutely needed.

The primary use-case for this is assigning appropriate values to items 8 through 11 of restart vector `RSEG`.  In turn, this will enable restoring these quantities from a restart file.

The implementation is based on reading the surrounding and calling source code, so I am not fully convinced that the way I am indexing into `segRates()` and `segPress()` is correct.  Corrections is much appreciated.

---
**Note**

This pull request depends on OPM/opm-common#514 and must not be merged before that is in place.

GitPaean commented:
Thanks for the request. I review the PR later this week. Please let me know if it is more urgent.

bska commented:
> I will review the PR later this week.

Thanks.  There's no rush.  I need to finish the feature set of OPM/opm-common#514 before this becomes unblocked.

GitPaean commented on opm/autodiff/WellStateFullyImplicitBlackoil.hpp:
```diff
@@ -793,6 +799,49 @@ namespace Opm
+                                 rate[pu.phase_pos[Gas]]);
+            }
+
+            segRes.segNumber = segmentID + 1;
```
I am not saying this is wrong, because it really depends on how you use it. 

More like a reminder,  from the input side, `segment number` is more or less a ID number, which is not necessarily related to its storage index.  

But it really depends how to define and use it. Please refer to the function `segmentNumberToIndex()` in `WellSegments` for reference.

GitPaean replied:
> capture the reverse mapping (index -> segment number)

With index, you can find the `Segment`, `Segment` has the information of `segment number`.

GitPaean replied:
>  I'll either have to capture the reverse mapping (index -> segment number) as a data member of WellStateFullyImplicitBlackoil or change the interface of report() to somehow enable passing the WellSegments from the outside.

I think this is right.

bska replied:
> > 'll either have to capture the reverse mapping (index -> segment number) as a data member of `WellStateFullyImplicitBlackoil` or change the interface of `report()` to somehow enable passing the `WellSegments` from the outside.
> 
> I think this is right.

Okay, in that case I think I'll add the data member because it can be built as part of `initWellStateMSWell()` and therefore serve as a consistency check for whether or not that function has been called.

GitPaean replied:
> Okay, in that case I think I'll add the data member because it can be built as part of `initWellStateMSWell()` 

Sure. 

> therefore serve as a consistency check for whether or not that function has been called.

I am not sure what this means, but I do not think it matters much here.

bska replied:
> > therefore serve as a consistency check for whether or not [`initWellStateMSWell()`] has been called.
> 
> I am not sure what this means, but I do not think it matters much here.

It'll be a signal that we should not attempt to return segment-related vectors for anything other than the top segment unless the user specifies [`--use-multisegment-well`](https://github.com/OPM/opm-simulators/blob/33da0aea09158408563a84393400f08efd41c38a/opm/autodiff/BlackoilModelParametersEbos.hpp#L68).  Otherwise we'd be indexing out of bounds since the `nseg_` member is uninitialised in that case and my new [`numSegments(wellID)`](https://github.com/OPM/opm-simulators/blob/28cb22d5d08fab9ebfb9d5ad9613241cc1adc9b7/opm/autodiff/WellStateFullyImplicitBlackoil.hpp#L837-L844) function would return useless values.

GitPaean replied:
> Otherwise we'd be indexing out of bounds since the nseg_ member is uninitialised in that case and my new numSegments(wellID) function would return useless values.

You can do it in your way. If you think it is a good idea to add a trivial initialization of `nseg_` similar to `top_segment_index_` at the end of  function `void init()`.  Leaving `nseg_` initialized was a mistake.

GitPaean commented:
I do not see anything obviously wrong  in this PR, except a reminder about the usage of `segment number`, which is by definition is a `ID`. Only the top segment has the fixed `segment number` `1`, the `segment number`s of other segments can be pretty random. 

And also, the naming style in the WellState file is different from the rest part (or most of the rest part) of the file. If you would like, you can change the naming style to be consistent with  (most of) the rest part, but I do not insist.

bska commented on tests/test_wellstatefullyimplicitblackoil.cpp:
```diff
@@ -0,0 +1,286 @@
+} // Anonymous
+
+BOOST_AUTO_TEST_SUITE(Segment)
+
```
> maybe you can re-organize a little bit.

I'll try.

bska commented:
I have updated this PR with a couple of more commits to address the initial review comments.  Commit 
d14c008 ensures that the `nseg_` data member always has a plausible value and commit e75a4a2 introduces a new data member (`seg_number_`) that provides the reverse (linear index to user-assigned ID) mapping for segments in the case of multi-segmented wells.  Regular (non-segmented) wells have a single segment (the top segment) whose ID is always one (1).

I have updated the unit test accordingly.

GitPaean commented:
jenkins build this please

GitPaean commented:
Any explanation that only the legacy simulation failed?
```
mpi.compareECLFiles_flow_legacy+SPE1CASE2
mpi.compareECLFiles_flow_legacy+SPE1CASE2_2P
mpi.compareECLFiles_flow_sequential+SPE1CASE1
mpi.compareECLFiles_flow_legacy+SPE3CASE1
mpi.compareECLFiles_flow_legacy+SPE9_CP_SHORT
mpi.compareRestartedSim_flow_legacy+SPE1CASE2_ACTNUM
mpi.compareRestartedSim_flow_legacy+SPE9_CP_SHORT```

atgeirr commented:
> It looks like all the legacy version of flow failed due to segmentation fault.

Keep in mind that we are moving these to a separate repository after the release. I think that repository will stay at the release 2018.10 level, to keep workable versions accessible.

GitPaean commented:
> Keep in mind that we are moving these to a separate repository after the release. I think that repository will stay at the release 2018.10 level, to keep workable versions accessible.

If I understand this correctly, we can do the merging after the moving?  Because now, without extra work, jenkins will not be happy.

bska commented:
> It looks like all the legacy version of flow failed due to segmentation fault.

Right.  I didn't apply my "always initialise `nseg_` and `seg_number_`" fix to `initLegacy()`.  Working on it.

bska commented:
> I didn't apply my "always initialise `nseg_` and `seg_number_`" fix to `initLegacy()`. Working on it.

Commit 6624ae7 applies that construction to `initLegacy()` too.  I'll rerun Jenkins now.  I expect that the `3D_MSW` test will fail due to results differing in `RSEG`.  If that turns out to be true we should update the reference data.

bska commented:
jenkins build this please

GitPaean commented:
>  It really should have.

If time permits, can you investigate it a little bit just to make sure nothing fishy is left anymore so we can get it in?  I suppose you are one most familiar with all the related.

And also, I think we have two MSW related tests, the other one is a 2D case under folder `msw_2d_h`. It looks like this one also does not fail.

GitPaean commented:
jenkins build this please

GitPaean commented:
I fixed the compilation error on my branch. I will create a pull request against ` report-segment-res branch on bska/opm-simulators ` when it is done.

bska commented:
jenkins build this please

GitPaean commented:
> Okay, thanks. I think this should go in sooner or later, but I am a little concerned actually that the 3D_MSW test did not fail. It really should have.

I am testing this functionality. Hopefully, I can help to give some information regarding this.

bska commented:
> > Okay, thanks. I think this should go in sooner or later, but I am a little concerned actually that the 3D_MSW test did not fail. It really should have.
> 
> I am testing this functionality. Hopefully, I can help to give some information regarding this.

Much appreciated.  It's actually [failing](https://ci.opm-project.org/job/opm-simulators-PR-builder/1396/testReport/junit/(root)/mpi/compareECLFiles_flow_3D_MSW/) now though, and for the right reasons (`RSEG` vector different).

GitPaean commented:
What is `RSEG` here?

bska commented:
> What is `RSEG` here?

Double precision quantities (e.g., pressures and flow rates) per well segment.

GitPaean commented:
I have been testing this PR. For the well data related to MS wells, it writes the segment data to the RESTART file and it recovers the segment data when restarting. 

However, it looks like some information related to oil phase (can be due to preprocessing, besides the function `processRestartSaturations_` from https://github.com/OPM/ewoms/pull/465/files) is not perfectly recover when restarting due to unclear causes, which prevents us from getting `perfect `restart running, which I have been looking for to help my debugging for big models. 

Anyway, this PR provides its intended functionality in a reasonable way along with OPM/opm-common#514 , so they should go in.

GitPaean commented:
jenkins build this opm-common=514 please

atgeirr commented on opm/autodiff/BlackoilWellModel_impl.hpp:
```diff
@@ -1724,6 +1723,85 @@ namespace Opm {
+    BlackoilWellModel<TypeTag>::
+    anyMSWellOpen(const Wells* wells, const int report_step) const
+    {
+        bool any_ms_well_open = false;
```
Would it not be easier to just loop over wells_ecl_ and call `isMultiSegment()` and `getStatus(report_step)` for each well? Or would that give something different?

GitPaean replied:
You are probably right. And also, the current approach did not consider whether it is parallel running or not, while WellState is distributed over processes. 

The difference might be related to whether it is parallel or not. Anyway, I will update it when I figure out the most proper way to do this.

GitPaean replied:
For now, looping through `wells` is still a good idea, although it does not make much difference in term of simulation. But `anyMsWellOpen` should probably be called `andMSWellOpenLocal` since it is local within the process.

atgeirr commented on opm/autodiff/BlackoilWellModel_impl.hpp:
```diff
@@ -1724,6 +1723,85 @@ namespace Opm {
+        // finding the location of the well in wells_ecl
+        const int nw_wells_ecl = wells_ecl_.size();
+        int index_well = 0;
+        for (; index_well < nw_wells_ecl; ++index_well) {
```
Looks ripe for `std::find()` and a lambda? Or is it too messy here?

GitPaean replied:
yes. I can do that. will update together with the comment above.

GitPaean commented:
There is some update is required on the opm-common, @bska has not opened pull request for it. 

It is not required for compilation (probably..), while it is required for correct behavoir.  Will address the comment later today.

bska commented:
> There is some update is required on the opm-common, @bska has not opened pull request for it.

I've created PR OPM/opm-common#709 to address the incorrect segment result loading.

bska commented:
jenkins build this opm-common=709 please

GitPaean commented:
jenkins build this opm-common=709 update_data please

bska commented:
The failed [test run 1419](https://ci.opm-project.org/job/opm-simulators-PR-builder/1419/testReport/junit/(root)/mpi/compareECLFiles_flow_MSW_MODEL_1/) shows among other differences:
```
Deviation results for keyword RSEG of type DOUB:
Average absolute deviation = 1.23032e+21
```
I've seen similar symptoms before, and it's **probably** a gas flow rate that's gone awry somewhere&mdash;either in the simulator or in the output.  Still, this value is so large that I'm hesitant to merge this PR.

GitPaean commented:
>  and it's probably a gas flow rate that's gone awry somewhere

`1.23e+21` is a little too big to be a flow rate if the simulation result is normal.

GitPaean commented:
More clues is the following, 

```
1 keyword exhibit failures
	RSEG
		Fails for 402 entries
		Largest absolute error: 1.2718329e+24
		Largest relative error: 1.0000000e+00
=== Executing comparison for init file ===
```

The `1.23e+21` average deviation is probably from this single value.  In anyway, one result has an abnormal big value and the other one is 0 there.

bska commented:
> > and it's probably a gas flow rate that's gone awry somewhere
> 
> `1.23e+21` is a little too big to be a flow rate if the simulation result is normal.

I absolutely agree.  Nevertheless, here's Item 9 of `RSEG` from Flow's [reference solutions](https://github.com/OPM/opm-tests/tree/aff36efe2d7929fd2131e45ca8fbbe6ffcbb336c/model1/opm-simulation-reference/flow) at the last time step for the three segmented wells of this model.  [Item 9](https://github.com/OPM/opm-common/blob/0d172c9a1c8730d802660bacccbb2aaad71a187e/src/opm/output/eclipse/AggregateMSWData.cpp#L549) is a weighted sum of surface flow rates for through the corresponding segment.

Segment ID | PROD-2 | PROD-3 | PROD-4
--------|-------------|----------|-----------
1 | 1.2718e+24 | 1.8133e+22 | 1.2389e+18
2 | 1.2718e+24 | 1.8133e+22 | 1.2389e+18
3 | 6.3592e+23 | 1.4506e+22 | 8.2594e+17
4 | N/A | 1.0880e+22 | 4.1297e+17
5 | N/A | 7.2532e+21 | N/A
6 | N/A | 3.6266e+21 | N/A

These values are much too large for a reasonable simulation result so we need to narrow down their origins.  I can think of three possible sources

1. Mistakes in unit conversion on output (output code)
2. Mistakes in combining flow rates to totals (output code)
3. The simulator actually creates those rate values (simulator/well model).

---

Finally, here are those summed flow rates above broken down as phase flow rates.  Note that gas rates being larger than the combined rate is expected, because we divide by 1000 when accumulating the gas rate into the total rate.  I think the oil and water rates look reasonable, but the gas rate is not in any way what you would expect.
#### Oil (SOFR, cubic metres per day) ####
Segment ID | PROD-2 | PROD-3 | PROD-4
--------|-------------|----------|-----------
1 | 0 | 0 | 8.2528e+02
2 | 0 | 0 | 8.2528e+02
3 | 0 | 0 | 5.5019e+02
4 | N/A | 0 | 2.7509e+02
5 | N/A | 0 | N/A
6 | N/A | 0 | N/A

#### Water (SWFR, cubic metres per day) ####
Segment ID | PROD-2 | PROD-3 | PROD-4
--------|-------------|----------|-----------
1 | 5.7517e+02 | 8.1495e+02 | 5.5671e+02
2 | 5.7517e+02 | 8.1495e+02 | 5.5671e+02
3 | 2.8758e+02 | 6.5196e+02 | 3.7114e+02
4 | N/A | 4.8897e+02 | 1.8557e+02
5 | N/A | 3.2598e+02 | N/A
6 | N/A | 1.6299e+02 | N/A

#### Gas (SGFR, cubic metres per day) ####
Segment ID | PROD-2 | PROD-3 | PROD-4
--------|-------------|----------|-----------
1 | 1.2718e+27 | 1.8133e+25 | 1.2389e+21
2 | 1.2718e+27 | 1.8133e+25 | 1.2389e+21
3 | 6.3592e+26 | 1.4506e+25 | 8.2594e+20
4 | N/A | 1.0880e+25 | 4.1297e+20
5 | N/A | 7.2532e+24 | N/A
6 | N/A | 3.6266e+24 | N/A

bska commented:
Update:

I don't think the high flow rates is a segment-related issue at all.  Rather, upon closer inspection, it appears to be a connection-related issue.  Running `MSW_MODEL1.DATA` with `--use-multisegment-well=true`, I get the following combined flow rates at the last time step:

Segment ID | PROD-2 | PROD-3 | PROD-4
--------|-------------|----------|-----------
1 | 1.3817e+03 | 9.6913e+02 | 1.3118e+03
2 | 1.3848e+03 | 9.7180e+02 | 1.3140e+03
3 | 7.0634e+02 | 4.5427e+02 | 6.6853e+02
4 | N/A | 4.5857e+02 | 6.3772e+01
5 | N/A | 4.6383e+02 | N/A
6 | N/A | 4.7044e+02 | N/A

Split into phase rates we get the following breakdowns:

#### SOFR ####
Segment ID | PROD-2 | PROD-3 | PROD-4
--------|-------------|----------|-----------
1 | 1.3817e+03 | 9.6913e+02 | 1.3118e+03
2 | 1.3848e+03 | 9.7180e+02 | 1.3140e+03
3 | 7.0634e+02 | 4.5427e+02 | 6.6853e+02
4 | N/A | 4.5857e+02 | 6.3772e+01
5 | N/A | 4.6383e+02 | N/A
6 | N/A | 4.7044e+02 | N/A

#### SWFR ####
Segment ID | PROD-2 | PROD-3 | PROD-4
--------|-------------|----------|-----------
1 | 7.5877e+00 | 4.1389e+01 | 1.5308e+02
2 | 7.5876e+00 | 4.1389e+01 | 1.5307e+02
3 | 1.3603e+00 | 4.1389e+01 | 1.5101e+02
4 | N/A | 4.1780e+01 | 1.4852e+02
5 | N/A | 4.2257e+01 | N/A
6 | N/A | 4.2859e+01 | N/A

#### SGFR ####
Segment ID | PROD-2 | PROD-3 | PROD-4
--------|-------------|----------|-----------
1 | 1.8595e+05 | 5.2460e+05 | 1.6229e+05
2 | 1.8595e+05 | 5.2459e+05 | 1.6229e+05
3 | 9.4725e+04 | 5.5490e+04 | 8.1162e+04
4 | N/A | 5.6014e+04 | 6.0236e+03
5 | N/A | 5.6655e+04 | N/A
6 | N/A | 5.7461e+04 | N/A

These are all mostly reasonable surface flow rates in cubic metres per day.  The top segment (segment ID 1) should normally be at least as large as all other values, so there may still be an issue here.  Nevertheless the order of magnitude looks good to me.  On the other hand, the connection flow rates are more questionable.  Here's what I see in GDB&mdash;slightly edited to prune away unset values and values not pertaining to the current discussion such as pressure and temperature&mdash;at report step 5 for the connection flow rates of well PROD-2:
```
connections = std::__debug::vector of length 2, capacity 2 = {{
  static restart_size = 2,
  index = 8,
  rates = {
    mask = (Opm::data::Rates::opt::wat | Opm::data::Rates::opt::oil | Opm::data::Rates::opt::gas),
    wat = -0.0033285089124777665,
    oil = -0.0042887244444892178,
    gas = -7360144187.8936663,
    ...
  },
  ...
}, {
  static restart_size = 2,
  index = 14,
  rates = {
    mask = (Opm::data::Rates::opt::wat | Opm::data::Rates::opt::oil | Opm::data::Rates::opt::gas), 
    wat = -0.0033285089124777665,
    oil = -0.0042887244444892178,
    gas = -7360144187.8936663,
    ...
  },
  ...
}}
```
These are the untouched SI-unit (cubic metres per second) flow rate values that come out of [`WellStateFullyImplicitBlackOil::report()`](https://github.com/OPM/opm-simulators/blob/67015480ea0cd0764363bcabfe8bb4c60114b41d/opm/autodiff/WellStateFullyImplicitBlackoil.hpp#L483).  If we're running the simulator with `--use-multisegment-well=false`&mdash;the current default setting&mdash;then we calculate segment flow rates on output based on connection flow rates.  The large gas flow rates at the connection level therefore dominate all derived segment quantities.

I now think that there's nothing wrong with the current PR (#1592) and that we should consider merging it into master accompanied by OPM/opm-common#709.  That said, we do need to get to the bottom of the connection gas flow rate issue here.

GitPaean commented:
> I now think that there's nothing wrong with the current PR (#1592) and that we should consider merging it into master accompanied by OPM/opm-common#709. That said, we do need to get to the bottom of the connection gas flow rate issue here.

That is good. I did not think we have abnormal simulation results.  I will address the comments and push the update shortly.

GitPaean commented:
It is very possible the connection flow phase rate is not updated properly at all for multisegment well model. I will check whether they are needed in the calculation. If the connection flow phase rate is used to calculate anything, like pressure drop, then it is a bug that impacts the simualtion. If not, it explains why we have not get hit by it yet.

bska commented:
> It is very possible the connection flow phase rate is not updated properly at all for multisegment well model. I will check whether they are needed in the calculation.

Just note that even if those connection rates aren't needed by the well model, we really should output reasonable values to the restart file's `XCON` vector (items 1, 2, and 3).  Those, in turn, are very useful for doing flow diagnostics analysis on simulations run by Flow.

GitPaean commented:
> Just note that even if those connection rates aren't needed by the well model, we really should output reasonable values to the restart file's XCON vector (items 1, 2, and 3). Those, in turn, are very useful for doing flow diagnostics analysis on simulations run by Flow.

Yes, you are right.  It needs to be fixed anyway.

GitPaean commented:
jenkins build this opm-common=709 opm-tests=106 please

GitPaean commented:
I addressed the comment. I think this PR should go in with OPM/opm-common#709

atgeirr commented:
Thanks for your work, merging!
